Add docs about how to resize volumes in GSP (example: Concourse Workers) (#345)
alphagov/gsp: docs/gds-supported-platform/resizing-persistent-volumes.md

# Resizing persistent volumes

The volumes you created in the beginning may not be big enough for
your ongoing needs.

As an example, we're going to double the space on the Concourse Worker
persistent volumes - from 64GiB to 128GiB. Replace
`gsp-concourse-worker` with the name of the resource you want to
resize the volume for.

1. Make a change to `charts/gsp-system/values.yml` with your desired
   sizing, like [in PR 343](https://github.com/alphagov/gsp/pull/343):

```yaml
...
persistence:
  worker:
    size: 128Gi
```

1. When merged, the PR will be released to sandbox. Changes must pass
   through sandbox via the full release process to make sure we haven't
   broken anything before going to the Verify cluster.

1. Pods in StatefulSets each have a volume. For Kubernetes to be able
   to resize the volumes, we need to detach them from the pods, by
   destroying the StatefulSet. To destroy the StatefulSet, run:

   ```sh
   $ gds sandbox kubectl -n gsp-system delete statefulset --cascade=false gsp-concourse-worker
   ```

>    You should destroy the StatefulSet when the Concourse release pipeline starts to fail. You will see an error message similar to:

Reviewer: It feels to tell someone *when* to run a command after you've already told them to run it.  I can imagine going through this document, running the `delete statefulset` line, and then reading this line and thinking "wait why didn't you tell me before"

Reviewer 2: Fair point. If you move this to before "To destroy the StatefulSet, run:..."  that should be ok hopefully. It pushes the action further away from the start of the numbered step, but you do need the context.

    > The StatefulSet “gsp-concourse-worker” is invalid: spec:
    > Forbidden: updates to statefulset spec for fields other than
    > ‘replicas’, ‘template’, and ‘updateStrategy’ are forbidden

   The `--cascade=false` flag makes sure that you do not delete the pods inside the StatefulSet.

   The `-deployer` pipeline will re-apply the StatefulSet. This will
   intentionally roll all the associated pods.

1. Once the pods are in a `Running` state, edit the

>    `PersisentVolumeClaim`s manually.

Reviewer: Is it worth saying *why* we have to do this?  If I've edited `persistence.worker.size` as described above, why do I also need to edit the PVCs?

Reviewer 3: I've added some context. Not sure how much detail we want to go into.

Reviewer: @blairboy362 I can't see what you changed?

Reviewer 3: I find it's much easier for others to see the changes when they've actually been pushed.

    - Find the name of the claims you want to delete. In this example,
      we want the three `concourse-worker`s.

      ```sh
      $ gds sandbox kubectl -n gsp-system get persistentvolumeclaim
      NAME                                        STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
      concourse-work-dir-gsp-concourse-worker-0   Bound    pvc-a633fb0a-76fe-11e9-bbf3-0ae9bf7a1ac0   64Gi      RWO            gp2            84d
      concourse-work-dir-gsp-concourse-worker-1   Bound    pvc-a6366add-76fe-11e9-bbf3-0ae9bf7a1ac0   64Gi      RWO            gp2            84d

>       concourse-work-dir-gsp-concourse-worker-2   Bound    pvc-a6386683-76fe-11e9-bbf3-0ae9bf7a1ac0   64Gi      RWO            gp2            84d

Reviewer: above, the text says `Pods in StatefulSets each have a volume`, but here it seems like all pods share the same volume?

Reviewer 3: The volume IDs are different...

Reviewer: oops, yes they are! 🙈

# Resizing persistent volumes

> The volumes you created in the beginning may not be big enough for

Reviewer 2: In the beginning of what? When you first create the cluster? Or something else?

Reviewer 3: When the thing using the persistent volume was first created...not sure how best to phrase that. Suggestions welcome!

Reviewer 2: I would rephrase to "You may need to resize your persistent volumes as your needs change."

your ongoing needs.

As an example, we're going to double the space on the Concourse Worker

> persistent volumes - from 64Gi to 128Gi. Replace

Reviewer 2: If gigabytes, use GB

Reviewer 3: Can we compromise on [GiB](https://www.iec.ch/si/binary.htm)?

Reviewer 2: Ok

`gsp-concourse-worker` with the name of the resource you want to
resize the volume for.

1. Make a change to `charts/gsp-system/values.yml` with your desired
   sizing, like [in PR 343](https://github.com/alphagov/gsp/pull/343):

```yaml
...
persistence:
  worker:
    size: 128Gi
```

> 1. When merged, the PR will be released to sandbox. Things must pass

Reviewer 2: Which things? Can we be more specific?

Reviewer 3: Changed to:
"Changes must pass through sandbox via the full release process..."

   through sandbox via the full resize process to make sure we haven't
   broken anything before going to the Verify cluster.

> 1. Pods in StatefulSets each have a volume. For Kubernetes to be able

Reviewer 2: is it StatefulSet, statefulset, or stateful set?

Reviewer 3: The k8s docs refer to StatefulSet pretty consistently, so we'll go with that!

   to resize the volumes, we need to detach them from the pods, by
   destroying the stateful set. The `--cascade=false` flag is
   important, as you don't want to delete the pods inside the stateful
   set.

   This step should be done as the Concourse release pipeline starts
   to fail with an error message like:

    > The StatefulSet “gsp-concourse-worker” is invalid: spec:
    > Forbidden: updates to statefulset spec for fields other than
    > ‘replicas’, ‘template’, and ‘updateStrategy’ are forbidden

   To destroy the StatefulSet, run:

    ```sh
    $ gds sandbox kubectl -n gsp-system delete statefulset --cascade=false gsp-concourse-worker
    ```

   The `-deployer` pipeline will re-apply the statefulset. This will
   intentionally roll all the associated pods.

> 

Reviewer 2: Content best practice is to put the action as close to the start of the numbered step as possible. The context is good, I would recommend rearranging the other content. For example:

"Pods in stateful sets each have a volume. For Kubernetes to be able to resize the volumes, we need to detach those volumes from the pods by destroying the stateful set. To destroy the stateful set, run:

    ```sh
    $ gds sandbox kubectl -n gsp-system delete statefulset --cascade=false gsp-concourse-worker
    ```
You should destroy the stateful set when the Concourse release pipeline starts to fail. You will see an error message similar to:

    > The StatefulSet “gsp-concourse-worker” is invalid: spec:
    > Forbidden: updates to statefulset spec for fields other than
    > ‘replicas’, ‘template’, and ‘updateStrategy’ are forbidden

The `--cascade=false` flag makes sure that you do not delete the pods inside the stateful set.

The `-deployer` pipeline re-applies the stateful set. This will intentionally roll all the associated pods.

Reviewer 2: Also, will the user know what "roll all the associated pods" means?

Reviewer 3: An SRE should definitely know what "roll the associated pods" means; we shouldn't be trying to turn this document into a "k8s fundamentals 101".

> 1. Once the pods are in a `Running` state, edit the

Reviewer 2: How do I know the pods are in a `Running` state?

Reviewer 3: Are you suggesting we include a command example here? As long as we don't have to explain all the options passed to it, shouldn't be too much of a digression.

Reviewer 2: Depends on if we are confident the user knows how to check this? If you think we need to tell people this then I would include it e.g.
" Check that the pods are in a `Running`state: 
```
CODE
```"
However, if this isnt necessary, then you can keep it as is.

Reviewer 3: I think the target audience for this document will know how to check. So I think I'll leave that for now.

    - Find the name of the claims you want to delete. In this example,
      we want the three `concourse-worker`s.

      ```sh
      $ gds sandbox kubectl -n gsp-system get persistentvolumeclaim
      NAME                                        STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
      concourse-work-dir-gsp-concourse-worker-0   Bound    pvc-a633fb0a-76fe-11e9-bbf3-0ae9bf7a1ac0   64Gi      RWO            gp2            84d
      concourse-work-dir-gsp-concourse-worker-1   Bound    pvc-a6366add-76fe-11e9-bbf3-0ae9bf7a1ac0   64Gi      RWO            gp2            84d
      concourse-work-dir-gsp-concourse-worker-2   Bound    pvc-a6386683-76fe-11e9-bbf3-0ae9bf7a1ac0   64Gi      RWO            gp2            84d
      ```

    - For each claim, run:

      ```sh
      $ gds sandbox kubectl -n gsp-system edit persistentvolumeclaim concourse-work-dir-gsp-concourse-worker-0
      $ gds sandbox kubectl -n gsp-system edit persistentvolumeclaim concourse-work-dir-gsp-concourse-worker-1
      $ gds sandbox kubectl -n gsp-system edit persistentvolumeclaim concourse-work-dir-gsp-concourse-worker-2
      ```

>       This will open `$EDITOR`. Find the line with the old volume size,

Reviewer 3: ```suggestion
      This will open `$EDITOR`. Update `.spec.resources.requests.storage` to the desired size, then save.
```

>       and change it to the desired size, then save.

Reviewer 3: ```suggestion
```
